### PR TITLE
Remove seasons from OS.

### PR DESCRIPTION
**Pull request by joseph-robertson**
Allow user-defined heating/cooling seasons (e.g., year-round for both) as an alternative to the automated HSP seasons.

### CODE REVIEW

File: measures/ResidentialAirflow/measure.rb [hunk 1]
```diff
@@ -2462,6 +2462,12 @@ def _processNaturalVentilationForUnit(model, runner, nat_vent, wind_speed, infil
       
     # Get heating and cooling seasons
     heating_season, cooling_season = HVAC.calc_heating_and_cooling_seasons(model, @weather, runner)
+    if building_unit.getFeatureAsString(Constants.SeasonHeating).is_initialized
+      heating_season = building_unit.getFeatureAsString(Constants.SeasonHeating).get.split(",").map {|i| i.to_f}
+    end
+    if building_unit.getFeatureAsString(Constants.SeasonCooling).is_initialized
+      cooling_season = building_unit.getFeatureAsString(Constants.SeasonCooling).get.split(",").map {|i| i.to_f}
+    end
```
Comment (shorowit) [inline measures/ResidentialAirflow/measure.rb]: @joseph-robertson @ejhw We should discuss how the natural ventilation seasons are impacted by this. If you choose to specify year-round heating and cooling, then the entire natural ventilation algorithm uses the "overlap" specification.

File: measures/ResidentialConstructionsWindows/measure.rb [hunk 2]
```diff
@@ -134,7 +134,24 @@ def run(model, runner, user_arguments)
...
+      break
+    end
+    
     heating_season, cooling_season = HVAC.calc_heating_and_cooling_seasons(model, weather, runner)
+    if building_unit.getFeatureAsString(Constants.SeasonHeating).is_initialized
+      heating_season = building_unit.getFeatureAsString(Constants.SeasonHeating).get.split(",").map {|i| i.to_f}
+    end
+    if building_unit.getFeatureAsString(Constants.SeasonCooling).is_initialized
+      cooling_season = building_unit.getFeatureAsString(Constants.SeasonCooling).get.split(",").map {|i| i.to_f}
+    end
```
Comment (shorowit) [inline measures/ResidentialConstructionsWindows/measure.rb]: @joseph-robertson @ejhw Similar comment for interior shading.

File: measures/ResidentialHVACSeasons/measure.rb [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    end
+    if cooling_start_month.is_initialized
+      cooling_start_month = month_map[cooling_start_month.get]
+    end
+    if cooling_end_month.is_initialized
+      cooling_end_month = month_map[cooling_end_month.get]
+    end    
+    
+    heating_season = Array.new(heating_start_month-1, 0) + Array.new(heating_end_month-heating_start_month+1, 1) + Array.new(12-heating_end_month, 0)
+    cooling_season = Array.new(cooling_start_month-1, 0) + Array.new(cooling_end_month-cooling_start_month+1, 1) + Array.new(12-cooling_end_month, 0)
```
Comment (shorowit) [inline measures/ResidentialHVACSeasons/measure.rb]: This logic isn't sufficient. A typical heating season would loop around the year, e.g., start_month=Oct and end_month=Mar. We should allow similar flexibility for cooling season (for the southern hemisphere). Please add unit tests that exercise this.

File: measures/ResidentialHVACSeasons/measure.rb [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+  end
+
+  # define the arguments that the user will input
+  def arguments(model)
+    args = OpenStudio::Measure::OSArgumentVector.new
+
+    #make a bool argument for using hsp seasons or not
+    arg = OpenStudio::Measure::OSArgument::makeBoolArgument("use_hsp_seasons", true)
+    arg.setDisplayName("Use HSP Seasons")
+    arg.setDescription("When the insulation does not completely fill the depth of the cavity, air film resistances are added to the insulation R-value.")
```
Comment (shorowit) [inline measures/ResidentialHVACSeasons/measure.rb]: The description needs to be updated. Also add some language the explains that the Start/End month arguments will be ignored if this is set to true.

**Comment by joseph-robertson** [comment]
@shorowit To have interior shading based on user-defined seasons, the window construction measure will have to come after the setpoint measures.

**Comment by joseph-robertson** [comment]
We've decided to default interior shading and natural ventilation to operate on the HSP only. Only the setpoint measures have the user-defined option.

File: measures/ResidentialHVACSizing/resources/hvac.rb [hunk 5]
```diff
@@ -1010,4 +1010,57 @@ def self.calc_heating_and_cooling_seasons(model, weather, runner=nil)
         return heating_season, cooling_season
     end
     
+    # Determine the heating/cooling seasons from the setpoint schedules
+    def self.get_heating_and_cooling_seasons(model, weather, runner=nil)
```
Comment (shorowit) [inline measures/ResidentialHVACSizing/resources/hvac.rb]: It looks like this method is no longer used and can be removed?

File: measures/ResidentialConstructionsWindows/measure.rb [hunk 6]
```diff
@@ -134,6 +134,17 @@ def run(model, runner, user_arguments)
...
+    units = Geometry.get_building_units(model, runner)
+    if units.nil?
+        return false
+    end
+    building_unit = nil
+    units.each do |unit|
+      building_unit = unit
+      break
+    end
+
```
Comment (shorowit) [inline measures/ResidentialConstructionsWindows/measure.rb]: This code doesn't seem used either.

File: measures/ResidentialHVACCoolingSetpoints/measure.rb [hunk 7]
```diff
@@ -50,6 +50,40 @@ def arguments(model)
     clg_wked.setDefaultValue("76")
     args << clg_wked    
     
+    #make a bool argument for using hsp seasons or not
+    arg = OpenStudio::Measure::OSArgument::makeBoolArgument("clg_use_hsp_seasons", true)
+    arg.setDisplayName("Use HSP Seasons")
+    arg.setDescription("Whether to use schedule values based on information contained in the EPW file or not. User-defined cooling start/end months will be ignored if this option is selected.")
+    arg.setDefaultValue(true)
+    args << arg
```
Comment (shorowit) [inline measures/ResidentialHVACCoolingSetpoints/measure.rb]: Some suggestions. Applies to heating setpoint measure too:
Arg: use_auto_season
Name: Use Auto Season
Description: "Specifies whether to automatically define the cooling season based on the weather file."
Default: false

File: measures/ResidentialHVACCoolingSetpoints/measure.rb [hunk 8]
```diff
@@ -50,6 +50,40 @@ def arguments(model)
...
+    month_display_names << "Aug"
+    month_display_names << "Sep"
+    month_display_names << "Oct"
+    month_display_names << "Nov"
+    month_display_names << "Dec"
+    
+    arg = OpenStudio::Measure::OSArgument::makeChoiceArgument("clg_start_month", month_display_names, false)
+    arg.setDisplayName("Cooling Start Month")
+    arg.setDescription("Start month of the cooling season.")
+    arg.setDefaultValue("Jun")
```
Comment (shorowit) [inline measures/ResidentialHVACCoolingSetpoints/measure.rb]: Set to Jan

File: measures/ResidentialHVACCoolingSetpoints/measure.rb [hunk 9]
```diff
@@ -50,6 +50,40 @@ def arguments(model)
...
+    args << arg
+    
+    arg = OpenStudio::Measure::OSArgument::makeChoiceArgument("clg_end_month", month_display_names, false)
+    arg.setDisplayName("Cooling End Month")
+    arg.setDescription("End month of the cooling season.")
+    arg.setDefaultValue("Sep")
```
Comment (shorowit) [inline measures/ResidentialHVACCoolingSetpoints/measure.rb]: Set to Dec

File: measures/ResidentialHVACCoolingSetpoints/measure.rb [hunk 10]
```diff
@@ -64,15 +98,45 @@ def run(model, runner, user_arguments)
...
+    # Get building units
+    units = Geometry.get_building_units(model, runner)
+    if units.nil?
+        return false
+    end
+    building_unit = nil
+    units.each do |unit|
+      building_unit = unit
+      break
+    end
```
Comment (shorowit) [inline measures/ResidentialHVACCoolingSetpoints/measure.rb]: This code isn't used.

File: measures/ResidentialHVACHeatingSetpoints/measure.rb [hunk 11]
```diff
@@ -64,15 +98,45 @@ def run(model, runner, user_arguments)
...
+    # Get building units
+    units = Geometry.get_building_units(model, runner)
+    if units.nil?
+        return false
+    end
+    building_unit = nil
+    units.each do |unit|
+      building_unit = unit
+      break
+    end
```
Comment (shorowit) [inline measures/ResidentialHVACHeatingSetpoints/measure.rb]: This code isn't used.

File: measures/ResidentialHVACHeatingSetpoints/measure.rb [hunk 12]
```diff
@@ -50,6 +50,40 @@ def arguments(model)
...
+    month_display_names << "Aug"
+    month_display_names << "Sep"
+    month_display_names << "Oct"
+    month_display_names << "Nov"
+    month_display_names << "Dec"
+    
+    arg = OpenStudio::Measure::OSArgument::makeChoiceArgument("htg_start_month", month_display_names, false)
+    arg.setDisplayName("Heating Start Month")
+    arg.setDescription("Start month of the heating season.")
+    arg.setDefaultValue("Oct")
```
Comment (shorowit) [inline measures/ResidentialHVACHeatingSetpoints/measure.rb]: Set to Jan

File: measures/ResidentialHVACHeatingSetpoints/measure.rb [hunk 13]
```diff
@@ -50,6 +50,40 @@ def arguments(model)
...
+    args << arg
+
+    arg = OpenStudio::Measure::OSArgument::makeChoiceArgument("htg_end_month", month_display_names, false)
+    arg.setDisplayName("Heating End Month")
+    arg.setDescription("End month of the heating season.")
+    arg.setDefaultValue("Mar")
```
Comment (shorowit) [inline measures/ResidentialHVACHeatingSetpoints/measure.rb]: Set to Dec

File: measures/ResidentialHVACHeatingSetpoints/measure.rb [hunk 14]
```diff
@@ -50,6 +50,40 @@ def arguments(model)
     htg_wked.setDefaultValue("71")
     args << htg_wked
 
+    #make a bool argument for using hsp seasons or not
+    arg = OpenStudio::Measure::OSArgument::makeBoolArgument("htg_use_hsp_seasons", true)
+    arg.setDisplayName("Use HSP Seasons")
+    arg.setDescription("Whether to use schedule values based on information contained in the EPW file or not. User-defined heating start/end months will be ignored if this option is selected.")
+    arg.setDefaultValue(true)
+    args << arg
```
Comment (shorowit) [inline measures/ResidentialHVACHeatingSetpoints/measure.rb]: Make similar changes as cooling setpoint measure.

File: measures/ResidentialHVACHeatingSetpoints/tests/process_heating_setpoints_test.rb [hunk 15]
```diff
@@ -125,8 +125,28 @@ def test_multifamily_new_construction
...
+  end
+  
+  def test_custom_heating_season_loop_around_year
+    args_hash = {}
+    args_hash["htg_use_hsp_seasons"] = "false"
+    expected_num_del_objects = {}
+    expected_num_new_objects = {"ScheduleRule"=>36, "ScheduleRuleset"=>3, "ThermostatSetpointDualSetpoint"=>1}
+    expected_values = {"heating_setpoint_sch_heating_season"=>71, "heating_setpoint_sch_overlap_season"=>71, "cooling_setpoint_sch_cooling_season"=>18000, "cooling_setpoint_sch_overlap_season"=>18000}
+    _test_measure("SFD_2000sqft_2story_SL_UA_3Beds_2Baths_Denver_Furnace_NoSetpoints.osm", args_hash, expected_num_del_objects, expected_num_new_objects, expected_values, 5)
+  end
```
Comment (shorowit) [inline measures/ResidentialHVACHeatingSetpoints/tests/process_heating_setpoints_test.rb]: Thanks!

**Pull request merged**